src/test/java/picard/sam/AbstractAlignmentMergerTest.java

```diff
@@ -192,24 +228,198 @@
 
                  */
 
-                {110, 105, 96, "12S80M18S", "13S97M", false, true, 105, 105, "12S80M18S", "22S88M", false,
+                {110, false, 105, 96, "12S80M18S", "13S97M", false, true, 105, 105, "12S80M18S", "22S88M",
                         default110LongR1Bases, default110LongR2Bases, default110LongR1Bases, default110LongR2Bases, null, null,
                         default110LongR1BaseQualities, default110LongR2BaseQualities, default110LongR1BaseQualities, default110LongR2BaseQualities, null, null},
 
-                {110, 105, 96, "12S80M18S", "13S97M", false, true, 105, 105, "12S80M8S10H", "10H12S88M", true,
+                {110, true, 105, 96, "12S80M18S", "13S97M", false, true, 105, 105, "12S80M8S10H", "10H12S88M",
                         default110LongR1Bases, default110LongR2Bases, sharedBases, sharedBases, default110LongR1ClippedBases, default110LongR2ClippedBases,
                         default110LongR1BaseQualities, default110LongR2BaseQualities, sharedQualities, sharedQualities, r1ClippedQualities10, r2ClippedQualities10},
+
+                // 123456789.123456789.123456789.123456789.123456789.123456789.123456789.123456789
+                //                  MMMMMMMMMMMMMMMMMMsssssssss>
+                //     <ssssssssMMM-MMMMMMMMMMMMMMMM
+                //          should become
+                // 123456789.123456789.123456789.123456789.123456789.123456789.123456789.123456789
+                //                  MMMMMMMMMMMMMMMMSSsssssssss>
+                //      <ssssssssSSSMMMMMMMMMMMMMMMM
+
+
+                {27, false, 18, 14, "18M9S", "8S3M1D16M", false, true, 18, 18, "16M11S", "11S16M",
```

Comment:
Could you add tests for the hardclipping case for these examples involving insertions and deletions as well?  Based on following the code and a simple test I tried out it seems to give an answer that I would considered "correct", but it gets there in a pretty unsafe round-about sort of way.  I suspect if it's not well tested the harclipping behavior in these cases could easily be unintentionally and unknowingly changed as a side effect of some future pr.

Reply:
added.